src/Features/CSharp/Portable/Braces/CSharpBracesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,404 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.AddBraces
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpAddBracesDiagnosticAnalyzer : DiagnosticAnalyzer, IBuiltInAnalyzer
+    {
+        private static readonly LocalizableString s_addBracesLocalizableTitle =
+            new LocalizableResourceString(nameof(FeaturesResources.Add_braces), FeaturesResources.ResourceManager,
+                typeof(FeaturesResources));
+
+        private static readonly LocalizableString s_addBracesLocalizableMessage =
+            new LocalizableResourceString(nameof(WorkspacesResources.Add_braces_to_0_statement), WorkspacesResources.ResourceManager,
+                typeof(WorkspacesResources));
+
+        private static readonly LocalizableString s_removeBracesLocalizableTitle =
+            new LocalizableResourceString(nameof(FeaturesResources.Remove_braces), FeaturesResources.ResourceManager,
+                typeof(FeaturesResources));
+
+        private static readonly LocalizableString s_removeBracesLocalizableMessage =
+            new LocalizableResourceString(nameof(WorkspacesResources.Remove_braces_from_0_statement), WorkspacesResources.ResourceManager,
+                typeof(WorkspacesResources));
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(_noneAddBracesDiagnosticDescriptor,
+                                  _infoAddBracesDiagnosticDescriptor,
+                                  _warningAddBracesDiagnosticDescriptor,
+                                  _errorAddBracesDiagnosticDescriptor,
+                                  _noneRemoveBracesDiagnosticDescriptor,
+                                  _infoRemoveBracesDiagnosticDescriptor,
+                                  _warningRemoveBracesDiagnosticDescriptor,
+                                  _errorRemoveBracesDiagnosticDescriptor);
+
+        private readonly DiagnosticDescriptor _noneAddBracesDiagnosticDescriptor;
```

Comment:
Right - the idea is that this analyzer can create new descriptors on the fly with different severities, we just don't want to return non-hidden descriptors from `SupportedDiagnostics`, because doing so causes the analyzer to run on all closed files, which is a performance hit.